### DISCUSSION TITLE
Specific Memory Specification per process ?

### DISCUSSION THREAD

**Discussion by ElNiak** [shadow/shadow#3335] [Q&A] [answered]
Hello,

I was asking myself if it was possible to do specific memory allocation per process like for to enable The Sniper Attack ?

Thanks !

**Comment by stevenengler** [comment]
If you want to limit the amount of memory available to each process running in Shadow, Shadow unfortunately doesn't have this capability. You could maybe modify Shadow to run each virtual host's processes in [control groups](https://man7.org/linux/man-pages/man7/cgroups.7.html), which might allow you to effectively set per-host memory limits. But I'm not familiar enough with cgroups to know if this will work out.

There might be alternatives, such as intercepting memory-related syscalls (ex: mmap and brk) and tracking how much memory is given to each process, but I think this would be brittle and difficult to handle edge cases for.

**Comment by robgjansen** [comment, marked as answer]
@ElNiak what exactly are you trying to achieve? Are you trying to make a process allocate more memory than usual so that it would be killed by the Linux kernel OOM killer?

In [the Sniper Attack paper](https://www.robgjansen.com/publications/sniper-ndss2014.html), the evaluation was carried out only with modifications to the Tor source code. If you are trying to reproduce the Sniper Attack in Shadow, I don't think you would need any changes inside of Shadow itself. You might need changes in the version of Tor that you run inside of your Shadow simulations. though. Additionally, Tor now implements DoS defenses against such attacks, so if you are going to simulate this in Shadow you would also need to change the torrc files for the relays running in Shadow to disable their DoS defenses, which would not represent the real world Tor deployment.

You might also be interested in the follow-up [the Point Break paper](https://www.robgjansen.com/publications/pointbreak-sec2019.html), where we explored what kinds of attacks are still possible while keeping the DoS defenses in place.

**Reply by ElNiak** [reply]
Hello 👋

No I am not trying to reproduce this exploit directly but similar ones.

I am trying to generate attacks on protocol implementations inside shadow simulator so they are reproductible.

For PoC, I think I will use the approach you mention by modifying them and then I will see. 

Thanks a lot for the insights ! 

I will keep you up-to-date if I decide to modify shadow to be more flexible or not.

**Reply by ElNiak** [reply]
If you want more context, you can checkout my repository:
- https://github.com/ElNiak/PANTHER

A paper describing our approach using Shadow has been accepted and will be released soon.

**Reply by robgjansen** [reply]
> PANTHER combines advanced techniques in network protocol verification, integrating the Shadow network simulator with the Ivy formal verification tool. 

Looks like a cool project! A direct link to the Shadow website or Github repo from your README would be much appreciated :)

For example:
PANTHER combines advanced techniques in network protocol verification, integrating [the Shadow network simulator](https://github.com/shadow/shadow) with the Ivy formal verification tool.

**Reply by ElNiak** [reply]
Thanks !! Yeah it is pretty cool hehe

I just did it !

**Reply by ElNiak** [reply]
(Don't hesite to put a star or what so ever too)

PS: sorry to do my "bitch"

**Reply by sporksmith** [reply]
This looks really cool! This makes me nervous though:

> PANTHER can simulate advanced persistent threats (APTs) in network protocols. By leveraging the Shadow network simulator, PANTHER models and analyzes the behavior of APTs in a controlled environment,

Shadow is *not* designed to contain software that may try to escape shadow's control. See https://shadow.github.io/docs/guide/security.html

**Reply by ElNiak** [reply]
Hello @sporksmith, I understand your fear ! but no worry, we want to test "good"/valid implem like picoquic, nginx, etc against our own controlled formal APT attacker and thus it is more safe
